src/compiler/utilities.ts

```diff
@@ -5939,18 +5939,31 @@ namespace ts {
     export interface SymlinkCache {
         getSymlinkedDirectories(): ReadonlyESMap<Path, SymlinkedDirectory | false> | undefined;
         getSymlinkedFiles(): ReadonlyESMap<Path, string> | undefined;
+        getRealPathToSymlinks(): ReadonlyESMap<Path, Path[]> | undefined;
         setSymlinkedDirectory(path: Path, directory: SymlinkedDirectory | false): void;
         setSymlinkedFile(path: Path, real: string): void;
     }
 
     export function createSymlinkCache(): SymlinkCache {
         let symlinkedDirectories: ESMap<Path, SymlinkedDirectory | false> | undefined;
         let symlinkedFiles: ESMap<Path, string> | undefined;
+        let realPathToSymlinks: ESMap<Path, Path[]>;
         return {
             getSymlinkedFiles: () => symlinkedFiles,
             getSymlinkedDirectories: () => symlinkedDirectories,
+            getRealPathToSymlinks: () => realPathToSymlinks,
             setSymlinkedFile: (path, real) => (symlinkedFiles || (symlinkedFiles = new Map())).set(path, real),
-            setSymlinkedDirectory: (path, directory) => (symlinkedDirectories || (symlinkedDirectories = new Map())).set(path, directory),
+            setSymlinkedDirectory: (path, directory) => {
+                // Don't add false directories or directories that have already been added
+                if (directory !== false && symlinkedDirectories?.has(path) !== true) {
+                    if (realPathToSymlinks === undefined) {
+                        realPathToSymlinks = new Map();
+                    }
+                    realPathToSymlinks.get(directory.realPath) === undefined ? realPathToSymlinks.set(directory.realPath, [path]) : realPathToSymlinks.get(directory.realPath)?.push(path);
+                }
+
+                (symlinkedDirectories || (symlinkedDirectories = new Map())).set(path, directory);

```

Comment:
We could move `(symlinkedDirectories || (symlinkedDirectories = new Map()))` but setting the path would be a little trickier since `symlinkedDirectories?.has(path)` will always return true.

Without the check for duplicates, our testing check was around 7 seconds instead of two

Reply:
Ah, that makes sense, let's keep it the way it is.